pkg/lib/cockpit-components-empty-state.tsx

```diff
@@ -0,0 +1,97 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import React from "react";
+import PropTypes from 'prop-types';
+import { Button, ButtonProps } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import {
+    EmptyStateActions, EmptyState, EmptyStateBody, EmptyStateFooter, EmptyStateHeader, EmptyStateIcon,
+    EmptyStateVariant, EmptyStateIconProps, EmptyStateHeaderProps
+} from "@patternfly/react-core/dist/esm/components/EmptyState/index.js";
+import { Spinner } from "@patternfly/react-core/dist/esm/components/Spinner/index.js";
+import "./cockpit-components-empty-state.css";
+
+export const EmptyStatePanel = ({
+    title,
+    paragraph,
+    loading = false,
+    icon,
+    action,
+    isActionInProgress = false,
+    onAction,
+    actionVariant = "primary",
+    secondary,
+    headingLevel = "h1"
+}: {
+    title?: EmptyStateHeaderProps["titleText"],
+    paragraph?: React.ReactNode,
+    loading?: boolean,
+    icon?: EmptyStateIconProps["icon"],
+    action?: React.ReactNode,
+    isActionInProgress?: boolean,
+    onAction?: ButtonProps["onClick"],
+    actionVariant?: ButtonProps["variant"],
+    secondary?: React.ReactNode,
+    headingLevel?: EmptyStateHeaderProps['headingLevel'],
+}) => {
+    const slimType = title || paragraph ? "" : "slim";
+    const iconFinal: EmptyStateIconProps["icon"] = icon ?? Spinner;
+
+    const emptyStateHeader = (loading || icon)
+        ? <EmptyStateHeader titleText={title}
+            headingLevel={headingLevel}
+            icon={<EmptyStateIcon icon={iconFinal} />}
+        />
+        : <EmptyStateHeader titleText={title}
+            headingLevel={headingLevel}
+        />;
```

Comment:
I think this can be simplified into:

```
 <EmptyStateHeader titleText={title}
            headingLevel={headingLevel}
            {... (loading || icon) && { icon: <EmptyStateIcon icon={icon ?? Spinner} /> }}
             />
```

Reply:
No hold the phone, this should be just?

```
                icon={(loading || icon) && <EmptyStateIcon icon={loading ? Spinner : icon} />} />
```

I have no idea why it is complicated in TypeScript

Reply:
Yeah this is where I'm fighting typescript and `exactOptionalPropertyTypes`. The conditions are perhaps too complicated and it complains that `icon` in `loading ? Spinner : icon` can be undefined... and because of the `exactOptionalPropertyTypes` check it's not not possible to pass undefined to an optional property